### ISSUE TITLE
question about unit ball surface features Visualization

### ISSUE THREAD

**Issue by SunJ1025** [John-520/DMsrTTLN#1]
Hello Dr.Lu,

First, thank you for your Nice work.

I'm now following your work and doing some futher research, I have a question that how can I get the Visualization results of unit ball surface features in Figure 6? Can you give me some futher explaination or is it convenient for you to share the code?

Thanks again for your reply!

**Comment by John-520** [comment]
> Hello Dr.Lu,
> 
> First, thank you for your Nice work.
> 
> I'm now following your work and doing some futher research, I have a question that how can I get the Visualization results of unit ball surface features in Figure 6? Can you give me some futher explaination or is it convenient for you to share the code?
> 
> Thanks again for your reply!


Thank you for your attention. Due to the complexity of recent affairs, the relevant code will be reorganized and made public. Alternatively, you can also add me on WeChat and send it to you separately.

**Comment by John-520** [comment]
> Hello Dr.Lu,
> 
> First, thank you for your Nice work.
> 
> I'm now following your work and doing some futher research, I have a question that how can I get the Visualization results of unit ball surface features in Figure 6? Can you give me some futher explaination or is it convenient for you to share the code?
> 
> Thanks again for your reply!









                '画出特征的球形分布图'
                model = torch.load('model.pkl')
                model.eval()
                
                data_data, label_label = T_test_data, T_test_label
                outputs_1 = model.predict_fea(data_data.cuda())    #.detach().numpy() predict predict_fea


                import torch
                import matplotlib.pyplot as plt
                from mpl_toolkits.mplot3d import Axes3D
                import matplotlib.cm as cm
                import matplotlib.colors as mcolors
                from matplotlib.colors import ListedColormap
                
                
                data = outputs_1.detach().cpu()
                labels = label_label.detach().cpu()
                # 使用PCA进行降维，将256个特征降维为3个特征
                pca_result = torch.pca_lowrank(data, q=3)
                # 取降维后的前三个特征用于绘制三维图
                components = pca_result[2]
                reduced_data = data.matmul(components)  # 将数据投影到PCA的主要方向上
                
                
                # 将数据投影到单位球体上
                normalized_data = reduced_data / reduced_data.norm(dim=1, keepdim=True)
                
                
                # 创建三维图
                fig = plt.figure()
                ax = fig.add_subplot(111, projection='3d')
                
                
                # 计算球体的重心
                center = torch.tensor([0.0, 0.0, 0.0])
                # 绘制三维单位球体
                u = np.linspace(0, 2 * np.pi, 20)
                v = np.linspace(0, np.pi, 20)         # 50 调节经纬度的密度
                x = np.outer(np.cos(u), np.sin(v))
                y = np.outer(np.sin(u), np.sin(v))
                z = np.outer(np.ones(np.size(u)), np.cos(v))
                ax.plot_wireframe(x, y, z, color='black', alpha=0.3)  # 绘制球体的经纬线   lightgray
                

                # 假设有三种自定义颜色
                custom_colors = ['#FF0000', '#00FF00', '#0000FF']
                custom_cmap = ListedColormap(custom_colors)
                
                # 绘制散点图
                ax.scatter(normalized_data[:, 0], normalized_data[:, 1], normalized_data[:, 2], marker='o')
                scatter = ax.scatter(
                    normalized_data[:, 0], normalized_data[:, 1], normalized_data[:, 2],
                    c=labels,  # 使用标签进行颜色映射
                    # cmap='magma',  # 选择你喜欢的颜色映射   magma  viridis
                    cmap=custom_cmap,
                    label=labels,  # 添加 label 参数
                    
                    marker='o'
                )
                
                
                # 绘制连接球体重心和样本点的线条
                for i in range(len(normalized_data)):
                    line_data = np.vstack([center.numpy(), normalized_data[i].numpy()])
                    ax.plot(line_data[:, 0], line_data[:, 1], line_data[:, 2], color='lightsteelblue', linestyle='-', linewidth=0.5)
                
                
                
                # 设置坐标轴标签
                ax.set_xlabel('Feature 1')
                ax.set_ylabel('Feature 2')
                ax.set_zlabel('Feature 3')
                
                ax.grid(False)  # 关闭坐标轴的网格线
                # 删除刻度线
                ax.set_xticks([])
                ax.set_yticks([])
                ax.set_zticks([])
                ax.axis('off')
                
                # 删除坐标轴的标题
                ax.set_xlabel('')
                ax.set_ylabel('')
                ax.set_zlabel('')
                # 设置背景颜色为纯白色
                ax.set_facecolor('white')
                
                
                # 设置图形的投影为球体
                ax.set_box_aspect([1, 1, 1])  # 设置坐标轴的比例为1:1:1，使图形投影为球体
                
                # 添加图例
                legend_labels = ['Label 0', 'Label 1', 'Label 2']
        
                # ax.legend(handles=scatter.legend_elements()[0], labels=legend_labels)
                ax.legend(handles=scatter.legend_elements()[0], labels=legend_labels, bbox_to_anchor=(0.8, 1), loc='upper left')
                
                # 调整图形布局
                plt.subplots_adjust(right=0.7)

                plt.savefig('.//保存的图片//特征球体.png',dpi=500, bbox_inches='tight')
                # 显示图形
                plt.show()
